torch/csrc/jit/frontend/script_type_parser.cpp

```diff
@@ -24,8 +24,12 @@ std::string collectQualname(const Select& select) {
 } // namespace
 
 TypePtr ScriptTypeParser::subscriptToType(
-    const std::string& typeName,
+    std::string& typeName,
     const Subscript& subscript) const {
+  if (typeName.rfind("typing.", 0) == 0) {
+    size_t pos = 7; // length of "typing."
+    typeName = typeName.substr(pos);
+  }
```

Comment:
@ansley I don't think this is the right approach. The first example in the PR description shows that the script type parser (or some component upstream) can already resolve `typing_List` and `typing_Tuple` fine, the only issue arises when we specifically see `typing_Tuple[()]`. I'll dig in to see how those aliases are resolved currently

Reply:
@ansley the issue appears to be in the resolution callback stuff here:

https://github.com/pytorch/pytorch/blob/9bfc1c4e0ee68b07557dd1138b3a9a6a587fa8a1/torch/_jit_internal.py#L73


When it encountered the expression `typing_Tuple[()]`, the parsing rules seem to diverge from Python and it discards the `()` tuple literal. Then it tries to lookup an empty string on the module, returning None

Reply:
question in my mind: is there a reason we can't use script_type_parser there?


or alternatively, eval the expression in the module environment then throw ann_to_type at it: https://github.com/pytorch/pytorch/blob/9bfc1c4e0ee68b07557dd1138b3a9a6a587fa8a1/torch/jit/annotations.py#L295